server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -862,7 +941,7 @@ protected void writeIndexGen(final RepositoryData repositoryData, final long exp
         // write the index file
         final String indexBlob = INDEX_FILE_PREFIX + Long.toString(newGen);
         logger.debug("Repository [{}] writing new index generational blob [{}]", metadata.name(), indexBlob);
-        writeAtomic(indexBlob, BytesReference.bytes(repositoryData.snapshotsToXContent(XContentFactory.jsonBuilder())), true);
+        writeAtomic(indexBlob, BytesReference.bytes(repositoryData.snapshotsToXContent(XContentFactory.jsonBuilder(), version)), true);
```

Comment:
Can we assert that repositoryData.shardGenerations does not contain `NEW_SHARD_GEN` entries? `NEW_SHARD_GEN` should not be written to disk either?

Reply:
I would in fact write `NEW_SHARD_GEN` to disk. You'd want this in case a shard snapshot was started on a data node but the node left during snapshotting. 
If you write this information to disk, a subsequent snapshot operation knows to not try and list out the shard directory (as `null` would imply if we were to just prune these values before writing).
It's really hard to do this in subsequent snapshots in `SnapshotsService` when creating the snapshot otherwise since we don't track the shard count for each index (and can't easily do that either since `RepositoryData` works with index names for now).

Reply:
> It's really hard to do this in subsequent snapshots in SnapshotsService when creating the snapshot otherwise since we don't track the shard count for each index (and can't easily do that either since RepositoryData works with index names for now).

ok, thanks for the explanation. That's really unfortunate. Can we assert then that we're never writing out a shard state metadata file with a `_new` prefix?

Reply:
Added assertion :)